ext/crates/fp/src/vector/internal/mod.rs

```diff
@@ -0,0 +1,686 @@
+use std::cmp::Ordering;
+
+use super::generic::{FpVectorIterator, FpVectorNonZeroIteratorP, FpVectorP, SliceMutP, SliceP};
+use crate::{
+    constants,
+    limb::{self, Limb, LimbLength},
+    prime::ValidPrime,
+};
+
+mod impl_internal;
+
+pub trait InternalBaseVectorP<const P: u32>: Sized {
+    /// Returns a pointer to the allocation containing the actual data. This is a raw pointer and
+    /// does not take lifetimes into account. It is the responsibility of the caller to ensure that
+    /// the pointer is not dereferenced after the allocation is freed.
+    ///
+    /// We use a pointer instead of a slice because otherwise handling the lifetimes is a huge mess.
+    /// In practice it is almost always better to use [`InternalBaseVectorP::_limbs`] to manipulate
+    /// the underlying data.
+    fn _as_ptr(&self) -> *const Limb;
+
+    /// Returns a description of the vector as a [`LimbLength`]. See there for the available
+    /// information.
+    fn _len(&self) -> LimbLength<P>;
+
+    fn _limbs(&self) -> &[Limb] {
+        unsafe { std::slice::from_raw_parts(self._as_ptr(), self._len().limbs()) }
+    }
+
+    fn _prime(&self) -> ValidPrime {
+        ValidPrime::new(P)
+    }
+
+    fn _is_empty(&self) -> bool {
+        self._len().logical() == 0
+    }
+
+    fn _is_zero(&self) -> bool {
+        let limb_range = self._len().limb_range();
+        if limb_range.is_empty() {
+            return true;
+        }
+        let (min_mask, max_mask) = self._len().limb_masks();
+        if self._limbs()[limb_range.start] & min_mask != 0 {
+            return false;
+        }
+
+        let inner_range = self._len().limb_range_inner();
+        if !inner_range.is_empty() && self._limbs()[inner_range].iter().any(|&x| x != 0) {
```

Comment:
Is the first check necessary/helpful?

Reply:
I think it is, otherwise `self._limbs()[inner_range]` panics IIRC

Reply:
This seems to work fine: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=fn%20main()%20%7B%0A%20%20%20%20let%20x%20%3D%20%5B1%2C%202%2C%203%5D%3B%0A%20%20%20%20println!(%22%7B%3A%3F%7D%22%2C%20%26x%5B0..0%5D)%0A%7D

Reply:
You're right, originally it was guarding against ranges like `1..0`, but those only happen if `limb_range` is already empty, which we now check beforehand.